jetstream/jetstream.go

```diff
@@ -46,100 +55,192 @@ type (
 		ObjectStoreManager
 	}
 
+	// Publisher provides methods for publishing messages to a stream.
+	// It is available as a part of [JetStream] interface.
+	// The behavior of Publisher can be customized using [PublishOpt] options.
 	Publisher interface {
-		// Publish performs a synchronous publish to a stream and waits for ack from server
-		// It accepts subject name (which must be bound to a stream) and message data
+		// Publish performs a synchronous publish to a stream and waits for ack
+		// from server. It accepts subject name (which must be bound to a stream)
+		// and message payload.
 		Publish(ctx context.Context, subject string, payload []byte, opts ...PublishOpt) (*PubAck, error)
-		// PublishMsg performs a synchronous publish to a stream and waits for ack from server
-		// It accepts subject name (which must be bound to a stream) and nats.Message
+
+		// PublishMsg performs a synchronous publish to a stream and waits for
+		// ack from server. It accepts subject name (which must be bound to a
+		// stream) and nats.Message.
 		PublishMsg(ctx context.Context, msg *nats.Msg, opts ...PublishOpt) (*PubAck, error)
-		// PublishAsync performs an asynchronous publish to a stream and returns [PubAckFuture] interface
-		// It accepts subject name (which must be bound to a stream) and message data
+
+		// PublishAsync performs an asynchronous publish to a stream and returns
```

Comment:
Isn't it though? Unless you are invoking `PublishAsync` concurrently, the messages will be sent over connection in order, so should be ordered.

Reply:
They may be sent in order, does not mean the stream will store them all, so you send a million and then a leader change happen, you will drop some in the middle.  Now you can never get them in order again.

Reply:
Right, thanks. Rephrased to:
```go
		// PublishAsync performs a publish to a stream and returns
		// [PubAckFuture] interface, not blocking while waiting for an
		// acknowledgement. It accepts subject name (which must be bound to a
		// stream) and message payload.
		//
		// PublishAsync does not guarantee that the message has been
		// received by the server. It only guarantees that the message has been
		// sent to the server and thus messages can be stored in the stream
		// out of order in case of retries.
```